### PR TITLE
Add a warning about unsupported save files

### LINKED ISSUE #1623: Crash when loading a level after changing AI code
When loading a save file the game loads the stack frames of the currently running programs from the save, but it loads the AI code from the game or mod files. If the AI code has changed, the code and the stack do not match. In some cases this can cause a crash.

0. Here is an archive with a mod and a save where the below steps have already been done:  [crash.zip](https://github.com/colobot/colobot/files/13402785/crash.zip). Just install the mod and open the save <---- crash
1. Create a mod. Use this script as AI for a programmable CreateObject
```
extern void object::Bug()
{
	while(true) {
		message("start");
		wait(1);
		wait(1);
		wait(1);
		wait(1);
		wait(1);
		wait(1);
		wait(1);
		wait(1);
		wait(1);
		wait(1);
		wait(1);
	}
}
```
2. Start the level
3. save
4. Change the script to
```
extern void object::Bug()
{
	while(true) {
		message("start");
		wait(1);
	}
}
```
5. load <-- the game will crash

It is possible to fix this specific example by making CBotListInstr trust the stack a little less, but the solution scales poorly: every CBot instruction would have to fully distrust the stack contents because they may have been created by a different instruction entirely.

I think a better solution would be to compute some kind or hash or checksum ([CRC](https://en.wikipedia.org/wiki/Cyclic_redundancy_check)?) of the AI source code and only load the stack if it matches the value from the save file. Otherwise the safest option is to restart the AI code from the beginning - should work for most levels.

If someone wants to fix this, it may make sense to fix #614 at the same time

### PR DESCRIPTION
**Pull request by hexagonrecursion**
## Problem

* Fixes #1623
* Also addresses some concerns I have about a number of edge cases where public classes or public functions may cause a crash when loading a save if there were changes to the CBot API (even if the changes do not break backward compatibility).

In https://github.com/colobot/colobot/pull/1668#discussion_r1599331147 @Emxx52  told me "loading old save files is generally unsupported".

## Solution

1. Loading an unsuported crashsave or quicksave now behaves as if there is no save
2. The -loadsave option also does not recognise a save as valid unless its version is supported
3. Loading an unsupported save via the save list asks the player to confirm

## Note to self: after this is merged I should implement a fix for:
* https://github.com/colobot/colobot/issues/1624
* https://github.com/colobot/colobot/issues/935
* https://github.com/colobot/colobot/issues/1565

### CODE REVIEW

**Comment by hexagonrecursion** [comment]
* Blocked: waiting for #1781 to land due to merge conflicts

**Comment by hexagonrecursion** [comment]
Who should I ask to review this?

**Review by eschan145 — APPROVED** [review]

**Comment by tomaszkax86** [comment]
I'll review this today or tomorrow.

File: colobot-base/src/common/version.h.cmake [hunk 1]
```diff
@@ -13,5 +13,9 @@ namespace Version
 
     static inline constexpr std::string_view FULL_NAME = "Colobot: Gold Edition @COLOBOT_VERSION_FULL@";
     static inline constexpr std::string_view VERSION = "@CMAKE_PROJECT_VERSION@";
+    static inline constexpr int MAJOR = @CMAKE_PROJECT_VERSION_MAJOR@;
+    static inline constexpr int MINOR = @CMAKE_PROJECT_VERSION_MINOR@;
+    static inline constexpr int PATCH = @CMAKE_PROJECT_VERSION_PATCH@;
+    static inline constexpr int TWEAK = @CMAKE_PROJECT_VERSION_TWEAK@;
```
Comment (Emxx52) [inline colobot-base/src/common/version.h.cmake:19]: TWEAK should be removed, we will not be using it in future releases.

File: colobot-base/src/ui/screen/screen_io_read.cpp [hunk 2]
```diff
@@ -168,10 +171,22 @@ bool CScreenIORead::EventProcess(const Event &event)
...
+        if (!name.has_value()) return false;
+        if (IsVersionSaveSupported(name.value()))
         {
-            m_main->StopSuspend();
-            m_main->ChangePhase(PHASE_SIMUL);
+            doRead();
+        }
+        else
+        {
+            m_dialog->StartQuestion("Version unsupported. Open anyway?", false, false, false, doRead);
```
Comment (Emxx52) [inline colobot-base/src/ui/screen/screen_io_read.cpp:189]: This text should be translatable (see [restext.cpp](https://github.com/colobot/colobot/blob/dev/colobot-base/src/common/restext.cpp), `update-pot`).

**Review by Emxx52 — CHANGES_REQUESTED** [review]

File: colobot-base/src/common/version.h.cmake (hunk 1 above)
Comment (tomaszkax86) [reply colobot-base/src/common/version.h.cmake:19]: Correct, we had ideas for it but that's unnecessary work for little benefits. Top level project's tweak number will also be removed.

File: CMakeLists.txt [hunk 3]
```diff
@@ -7,6 +7,9 @@ cmake_minimum_required(VERSION 3.21)
 
 cmake_policy(SET CMP0025 NEW)
 
+# Loading save files created by a previous version of the game is
+# unsupported with one exception: the first three version components have to match
+# and the fourth version component has to be greater or equal.
```
Comment (tomaszkax86) [inline CMakeLists.txt:12]: This description has no place in top level CMakeLists.txt. It should be moved near relevant implementation details, like in a comment before the function that checks save file version.

File: colobot-base/src/ui/screen/screen_io_read.cpp [hunk 4]
```diff
@@ -168,10 +171,22 @@ bool CScreenIORead::EventProcess(const Event &event)
 
     if ( event.type == EVENT_INTERFACE_IOREAD )
     {
-        if(IOReadScene() && m_inSimulation)
+        auto doRead = [&]() {
```
Comment (tomaszkax86) [inline colobot-base/src/ui/screen/screen_io_read.cpp:174]: Minor: `{` should be in next line

File: colobot-base/src/ui/screen/screen_io.h [hunk 5]
```diff
@@ -41,6 +42,7 @@ class CScreenIO : public CScreen
     void IODeleteScene();
     void IOWriteScene();
     bool IOReadScene();
+    std::optional<std::filesystem::path> GetSceneName();
```
Comment (tomaszkax86) [inline colobot-base/src/ui/screen/screen_io.h:45]: Minor: if possible, make the method `const` qualified

File: colobot-base/src/ui/screen/screen_io_read.cpp (hunk 2 above)
Comment (tomaszkax86) [inline colobot-base/src/ui/screen/screen_io_read.cpp:189]: I'd change the text to be more descriptive. Something like `"This save is from a different version of the game and might not work correctly. Do you want to open it anyway?"`

**Review by tomaszkax86 — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 3 above)
Comment (hexagonrecursion) [reply CMakeLists.txt:12]: Removing TWEAK made this comment redundant

**Comment by hexagonrecursion** [comment]
@tomaszkax86 @Emxx52 Thanks. Done

**Comment by hexagonrecursion** [comment]
In case this somehow fell through the cracks: I have applied review suggestions. Can someone review this again please?

**Review by tomaszkax86 — APPROVED** [review]

**Review by Emxx52 — CHANGES_REQUESTED** [review]
Works well, now it's only missing updated po files. Please run `make update-pot` and then build the project to generate them (you might need to rebase first).

**Comment by hexagonrecursion** [comment]
> Works well, now it's only missing updated po files. Please run `make update-pot` and then build the project to generate them (you might need to rebase first).

Thanks. Done

**Review by TheGiraffe3 — APPROVED** [review]
waring in the title to warning?
Code changes look good.

**Comment by hexagonrecursion** [comment]
> waring in the title to warning?

I do not understand. Are you talking about the commit message? Are you talking about the title of the pull request? Are you talking about the text of the new pop-up I added?

Are you suggesting we should change it?... Oh. I see the typo. Thanks.

**Comment by hexagonrecursion** [comment]
@TheGiraffe3 in order to understand your comment I had to:
1. Notice that one of the words is spelled wrong
2. Somehow guess that your spelling is intentional

A more reliable way to communicate this would be: "There is a typo in the title: waring instead of warning"

**Comment by TheGiraffe3** [comment]
My apologies.

**Review by Emxx52 — APPROVED** [review]

**Pull request merged**